### PR TITLE
Mock MeterStatusAPI

### PR DESCRIPTION
**Pull request by SimonRichardson**
## Description of change

The following mocks out the meter status api. The tests mock out
the state type and consume only the methods that we require. The
new tests don't require the database when running "WatchMeterStatus".

Interestingly we require a new constructor so that we satisfy the
allfacades registry setup, as that actually tests the function
type is `(*state.State, facade.Resources, facade.Authorizer)`,
which is a shame, because that restricts what we can get the meter
status to consume.

## Information

https://discourse.jujucharms.com/t/state-watchers/217

### CODE REVIEW

File: apiserver/facades/agent/meterstatus/meterstatus.go [hunk 1]
```diff
@@ -73,6 +85,18 @@ func NewMeterStatusAPI(
 	}, nil
 }
 
+// NewMeterStatusAPIWithState creates a new MeterStatusAPI, but uses the
+// state.State for creating the MeterStatusAPI. This is so we fulfill the
+// allfacades registration setup (i.e. it requires the state rather than
+// the well defined interface that the MeterStatusAPI requires)
+func NewMeterStatusAPIWithState(
```
Comment (howbazaar) [inline apiserver/facades/agent/meterstatus/meterstatus.go]: Can we please not use this facade constructor?
Instead please use this form:
```go
func NewMeterStatusFacade(ctx facade.Context) (*MeterStatusAPI, error) {
```

Comment (SimonRichardson) [reply apiserver/facades/agent/meterstatus/meterstatus.go]: @howbazaar of course :dancer:

File: apiserver/facades/agent/meterstatus/meterstatus_test.go [hunk 2]
```diff
@@ -91,5 +94,113 @@ func (s *meterStatusSuite) TestGetMeterStatus(c *gc.C) {
...
+	result, err := status.WatchMeterStatus(args)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(result, gc.DeepEquals, params.NotifyWatchResults{
+		Results: []params.NotifyWatchResult{
+			{NotifyWatcherId: "1"},
+		},
+	})
+}
+
+func (s *meterStatusSuite) setupMeterStatusAPI(c *gc.C, fn func(*mocks.MockMeterStatusState, *mocks.MockResources, *mocks.MockAuthorizer)) (*meterstatus.MeterStatusAPI, *gomock.Controller) {
```
Comment (manadart) [inline apiserver/facades/agent/meterstatus/meterstatus_test.go]: This is subjective, but I would make a struct with the mocks as members. 

Then you can shorten up the signature on all the calls, plus you don't have to change them all if say another mock or testing artefact was added.

**Comment by SimonRichardson** [comment]
!!build!!

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**